Use push flyout for Discover document flyout

lukasolson opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/163798. 
Closes #160306

Uses `"push"` style for the document viewer flyout in Discover.

Before:

<img width="1552" alt="image" src="https://github.com/elastic/kibana/assets/1178348/7e3e7c35-7317-44ae-b614-55eb337f742b">

After:

<img width="1552" alt="image" src="https://github.com/elastic/kibana/assets/1178348/542b78e6-3c82-47a5-b35f-1c8635e21580">

### Checklist

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

lukasolson commented on src/plugins/discover/public/components/discover_grid_flyout/discover_grid_flyout.tsx:
```diff
@@ -251,6 +254,15 @@ export function DiscoverGridFlyout({
             textBasedHits={isPlainRecord ? hits : undefined}
           />
         </EuiFlyoutBody>
+        <EuiFlyoutFooter>
```
From the EUI docs (https://elastic.github.io/eui/#/layout/flyout#push-versus-overlay):

> Also, it is good to include a close button in the footer for a larger hit target than the small close button provides.

elasticmachine commented:
Pinging @elastic/kibana-data-discovery (Team:DataDiscovery)

kertal commented on src/plugins/discover/public/components/discover_grid_flyout/discover_grid_flyout.tsx:
```diff
@@ -183,6 +187,15 @@ export function DiscoverGridFlyout({
+          <EuiFlexGroup>
+            <EuiFlexItem grow={false}>
+              <EuiButtonEmpty iconType="cross" onClick={onClose} flush="left">
+                Close
```
yes, this make sense, one ask, pls add i18n for `Close` 🙏

kertal commented:
@elasticmachine merge upstream

davismcphee reviewed (commented):
Looking good overall! A couple point of feedback from testing:

Now that the sidebar is resizable and the doc flyout pushes the content, the sidebar and main panels can become squished when the flyout is open and the screen is on the smaller side of XL (so still a push flyout). It seems like reducing the min width of the main panel improves this. I'd recommend updating `discover_resizable_layout.tsx` to use `const minMainPanelWidth = euiTheme.base * 24;` instead of the current value.

I also noticed that since the flyout pushes the content when it opens, you can lose sight of the selected document row as the data grid resizes. I wonder if it's possible to do something like auto-scroll to the selected doc after opening the flyout?
![lose_doc](https://github.com/elastic/kibana/assets/25592674/6cad4de9-fae9-4b80-ba50-34f0ed2ddad2)

It also looks like after I close the push flyout, there are some rendering issues with the data grid. This is likely just a bug with `EuiDataGrid`, but I wonder if we can do something about it. It's possible that implementing some sort of auto-scroll functionality might force the data grid to render and fix this issue as well:
![render_issues](https://github.com/elastic/kibana/assets/25592674/5483c016-2b6c-4857-b2fa-c9c26d6b7496)

Otherwise I do find the flyout a bit wide overall. Even on a large screen it took up about half the space, but I think this could be reduced when we migrate to floating actions in the doc viewer table, so I wouldn't consider it blocking unless others feel strongly about it.

kertal commented:
I feel like, given there were new changes under the hood, like the sizable field list, and also the potential vanishing of the actually selected row when the autosized document column is rendered, it should be reviewed again for design changes, and in the context of our ongoing redesign FYI @andreadelrio @MichaelMarcialis @ninoslavmiskovic 
I feel like we should not rush in this significant change, here's a deployment to test:
[deployment](https://kertal-pr-166406-unified-doc-viewer-push-flyout.kbndev.co/tcj/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:60000),time:(from:now-15m,to:now))&_a=(columns:!(),filters:!(),index:ff959d40-b880-11e8-a6d9-e546fe2bba5f,interval:auto,query:(language:kuery,query:''),sort:!(!(order_date,desc))))

andreadelrio commented:
> I feel like, given there were new changes under the hood, like the sizable field list, and also the potential vanishing of the actually selected row when the autosized document column is rendered, it should be reviewed again for design changes, and in the context of our ongoing redesign FYI @andreadelrio @MichaelMarcialis @ninoslavmiskovic I feel like we should not rush in this significant change, here's a deployment to test: [deployment](https://kertal-pr-166406-unified-doc-viewer-push-flyout.kbndev.co/tcj/app/discover#/?_g=(filters:!(),refreshInterval:(pause:!t,value:60000),time:(from:now-15m,to:now))&_a=(columns:!(),filters:!(),index:ff959d40-b880-11e8-a6d9-e546fe2bba5f,interval:auto,query:(language:kuery,query:''),sort:!(!(order_date,desc))))

Seeing this in action, I think I would not implement this change until we implement #164617. The left white space for me is a blocker to implement this push flyout. When we move to floating actions we'll be able to reduce the width of this flyout. I also think that when we implement the push flyout it should come with the ability to drag-to-resize, similar to what we now have for the field list.

ninoslavmiskovic commented:
The vanishing row is a bug in my opinion. 

Also I did not expect the push fly out to be that "aggressive" on resizing. 

Could we e.g work with how it pushes the content or collapse the field list automatically? 

Agree on next iteration with working on the re-sizing, however before see it as a blocker. 

Is there anything we can work on how it renders ?

kertal commented on src/plugins/discover/public/components/discover_grid_flyout/discover_grid_flyout.tsx:
```diff
@@ -113,7 +115,9 @@ export function DiscoverGridFlyout({
+        pushMinBreakpoint="xl"
         data-test-subj="docTableDetailsFlyout"
         onKeyDown={onKeyDown}
         ownFocus={false}
```
I think `ownFocus` should be true now, this would not apply for the push, but for the overlay style
```suggestion
        ownFocus={true}
```

andreadelrio commented:
@lukasolson this is awesome! Design comments below:

1. I think we need to adjust the initial proportions when opening the flyout. The main table should be wider. 

#### Initial dimensions [Currently in this PR]:
<img width="1079" alt="image" src="https://github.com/elastic/kibana/assets/4016496/943c3280-cd43-43e6-86ff-9214da0cc3fc">

####  Initial dimensions [Suggestion]: We can aim for a point where the unified search bar isn't wrapping into a second line

- Design detail for this scenario: Adjust the KQL placeholder text so it doesn't get cut off (to `Filter your data`).
- When we implement #174745, I would consider reducing the default flyout width even further. 

<img width="1079" alt="image" src="https://github.com/elastic/kibana/assets/4016496/89955516-c048-43bd-8368-f6d85bdff291">

2. Can we store the width that the user sets? 
Scenario: _I expand a document and set the flyout width that works for me (400px). I close the flyout and continue exploring my data. Then, I decide to expand another document and I would like the flyout to already be in the width that I like (400px)._

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

lukasolson commented:
@andreadelrio Thanks for the feedback! I've addressed the first part of it.

> Can we store the width that the user sets?

Should be doable, but I've reached out to the EUI team to see if there's an event handler or something I can use to hook into that. I didn't see anything like we have on other resizable components.

stratoula commented on src/plugins/unified_search/public/query_string_input/query_string_input.tsx:
```diff
@@ -22,7 +22,6 @@ import {
   EuiTextArea,
   htmlIdGenerator,
   PopoverAnchorPosition,
-  toSentenceCase,
```
My dear Lukas, I am always fond of cleanups but how is this change related to the push flyout in Discover?

lukasolson replied:
Good question :) Up a little bit @andreadelrio had pointed out that sometimes the resizable flyout can cause the placeholder text in the search bar to wrap, and suggested we update it:

> - Design detail for this scenario: Adjust the KQL placeholder text so it doesn't get cut off (to `Filter your data`).

Happy to hear your input here as well!

stratoula replied:
I see, @andreadelrio the reason tha Caroline and I added the language to the placeholder is because there is no indication in the UI about which language is on. To see that a user should open the menu popover (but it is not obvious

<img width="705" alt="image" src="https://github.com/elastic/kibana/assets/17003240/3d2c65a0-8dbf-4a2c-bd5d-791d45db3c04">

So I propose:
- Change to a shorter text which included the language
- Don't do this change now, but take this under consideration to your unified search redesign

andreadelrio replied:
@lukasolson I was hoping we would only shorten the text when the screen wasn't wide enough to display the whole string. In any case, I hadn't considered the context @stratoula is mentioning (thanks!). I suggest we hold off on this change until we implement other changes to the unified search bar. We're considering exposing the language toggle in the main bar.

stratoula replied:
Yes this sounds better! Thanx Andrea and Lukas!

lukasolson replied:
:+1: Reverted in dfebc4061805a9874ac219255b5df1dae9ac2de9

kpollich reviewed (approved):
Fleet test change LGTM

stratoula commented:
/ci

kertal commented on src/plugins/discover/public/components/discover_grid_flyout/discover_grid_flyout.tsx:
```diff
@@ -251,7 +267,18 @@ export function DiscoverGridFlyout({
           )}
         </EuiFlyoutHeader>
         <EuiFlyoutBody>{bodyContent}</EuiFlyoutBody>
```
@lukasolson this comment it referring to the element being rendered in the `EuiFlyoutBody` 

let's bring this over the finishing line by adjusting the responsiveness of the action column of the . something like this should work with a change at the beginning of the  `DocViewerTable` component

https://github.com/elastic/kibana/blob/f7341700b9cbed2fd6c2fa623d189e214edf4aa4/src/plugins/unified_doc_viewer/public/components/doc_viewer_table/table.tsx#L111-L125

```TypeScript
  const { euiTheme } = useEuiTheme();
  const [ref, setRef] = useState<HTMLDivElement | HTMLSpanElement | null>(null);
  const dimensions = useResizeObserver(ref);
  const showActionsInsideTableCell = dimensions?.width
    ? dimensions.width > euiTheme.breakpoint.m
    : false;
```
    
 With  `setRef` being used the the `EuiFlexGroup` wrapper
 
 Then the action column would no longer be responsive to the window width, but to the flyout width, which is a better pattern anyway
 
 Thx for the feedback @andreadelrio and `useResizeObserver` hint @davismcphee

davismcphee replied:
One other note: we should also make the single document and surrounding documents buttons responsive based on the flyout width since they currently use the window width too. We should be able to use the same resize observer approach for both the buttons and table actions.

>Should we use `euiTheme.breakpoint.m` or `euiTheme.breakpoint.l`?

We could use a preset breakpoint, but it might even be better to just use `euiTheme.base * X` and get a value close to when the UI elements start to "break".

lukasolson commented:
@kertal @davismcphee I've updated the behavior here and was hoping you could take a look to see if it seems expected.

I kept the `isMobileScreen` detection because IMO it does make sense to switch to the drop-down actions when viewing on mobile. However I updated the `isLargeScreen` behavior to be based on the actual flyout size.

kertal commented:
> @kertal @davismcphee I've updated the behavior here and was hoping you could take a look to see if it seems expected.
> 
> I kept the `isMobileScreen` detection because IMO it does make sense to switch to the drop-down actions when viewing on mobile. However I updated the `isLargeScreen` behavior to be based on the actual flyout size.

👍 this all makes sense, maybe `isLargeScreen` could be relabeled to `isLargeFlyout` or `isLargeContainer` so it also makes sense outside of this context.

Btw, I took a brief look at the test failures and due to the push flyout the following log explorer test failed here:

https://github.com/elastic/kibana/blob/65b5298a85c2920fd479a02a5a9bcd9862f84f51/test/functional/services/data_grid.ts#L224-L230

When the flyout is opened for the second time, it seems the cell indicating the first column is no longer found? rowindex stays -1, and therefore the test fails ... this is pretty. pretty weird!

davismcphee commented on src/plugins/discover/public/components/discover_grid_flyout/discover_grid_flyout_actions.tsx:
```diff
@@ -34,11 +35,36 @@ export interface DiscoverGridFlyoutActionsProps {
+  const [ref, setRef] = useState<HTMLDivElement | HTMLSpanElement | null>(null);
+  const dimensions = useResizeObserver(ref);
+  const isMobileScreen = useIsWithinBreakpoints(['xs', 's']);
+  const isLargeScreen = dimensions?.width ? dimensions.width > euiTheme.base * 30 : false;
```
```suggestion
  const isLargeFlyout = dimensions?.width ? dimensions.width > euiTheme.base * 30 : false;
```
Nit: I agree with @kertal  that `isLargeFlyout` might be more appropriate for this.

davismcphee reviewed (commented):
The latest changes look good, and it seems to be working well! I left a couple of very minor nits, but nothing significant. I think we're pretty close to this being mergable IMO.

I did notice a couple of small UX things while testing that I think could be touched up, and since I was in there anyway I figured I might as well throw together a quick touchups PR against the branch to consider: https://github.com/lukasolson/kibana/pull/20.

andreadelrio reviewed (changes requested):
This is looking good @lukasolson! I have a few quick design suggestions:

1) Reduce the overall padding in the flyout (to `m` - 16px)
2) Reduce the heading size of the flyout
3) Get rid of the `<strong>` tag for the actions text in `DiscoverGridFlyoutActions`

This will achieve two things:
1) Free up some space to show more content
2) Match the style of the Lens inline editor flyout better

The result should be this:

<img width="1443" alt="image" src="https://github.com/elastic/kibana/assets/4016496/f61e490c-2930-4534-8000-bbe3f0968702">

lukasolson commented on x-pack/test/functional/apps/observability_logs_explorer/flyout.ts:
```diff
@@ -79,7 +79,9 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
       await dataGrid.closeFlyout();
+    });
 
+    it('should not display a log level badge when not available', async () => {
```
For some reason this test (and the others below) were failing because the web elements had become "stale", and were throwing errors. I'm not sure how the changes in this PR would affect this, but breaking these into separate `it()` statements seems to fix the issue and still test the same behavior.

tonyghiani replied:
[This PR](https://github.com/elastic/kibana/pull/180262) will actually remove most of these tests after moving the doc view implementation, I'll take care of the conflicts but this LGTM.

kertal reviewed (approved):
LGTM, looking great and working snappy 👍  
Tests locally with Safari, Firefox, Chrome, all fine and work as expected.
A note about the tests, that's pretty strange, it seems when using the push flyout our code in the test to get the row cells no longer works. Might be caused because when the push flyout is closed, this leads the table / layout to be pre-rendered. Not a blocker but we might need to investigate what's happening here , what also might have worked instead of creating separate tests, is just to not close the flout when showing another row, since this should work with the push flyout

davismcphee commented on test/accessibility/apps/discover.ts:
```diff
@@ -133,7 +133,8 @@ export default function ({ getService, getPageObjects }: FtrProviderContext) {
       if (await testSubjects.exists('openFieldActionsButton-Cancelled')) {
-        await testSubjects.click('openFieldActionsButton-Cancelled');
+        await testSubjects.click('openFieldActionsButton-Cancelled'); // Open the actions
+        await testSubjects.click('openFieldActionsButton-Cancelled'); // Close the actions
```
I think we should close the menu after we call `await a11y.testAppSnapshot();` to make sure our a11y snapshot includes the open menu.

davismcphee reviewed (commented):
Just about good to go on my end! A couple of minor points of feedback, but it's just about there IMO.

I did notice that my workaround to prevent breaking the Discover layout doesn't work when there's a nav sidebar like in serverless or the O11y solution nav, but I don't think there's much we can do about that unless we make all of Discover responsive based on container width, so I think it's good for now:
<img width="1997" alt="overlap" src="https://github.com/elastic/kibana/assets/25592674/9f491bfd-0e43-42cc-bb24-8ef101326a16">

andreadelrio reviewed (approved):
LGTM, thanks!

davismcphee reviewed (approved):
Looks like my remaining concerns are addressed, so in that case this LGTM 👍 Great work on this, and I'm excited to get it merged!

I noticed one minor issue where the Logs Explorer flyout default width isn't being set now, but that's actually unrelated to these changes and is a separate bug with the customization framework, so I'll open up a separate PR to address it.
